crates/wasmparser/src/validator/operators.rs

```diff
@@ -71,6 +59,34 @@ pub(crate) struct OperatorValidator {
     operands: Vec<Option<ValType>>,
 }
 
+// No science was performed in the creation of this number, feel free to change
+// it if you so like.
+const MAX_LOCALS_TO_TRACK: usize = 50;
+
+#[derive(Default)]
+struct Locals {
+    // Total number of locals in the function.
+    num_locals: u32,
+
+    // The first MAX_LOCALS_TO_TRACK locals in a function. This is used to
+    // optimize the theoretically common case where most functions don't have
+    // many locals and don't need a full binary search in the entire local space
+    // below.
+    first: Vec<ValType>,
```

Comment:
Technically, since `ValType` is a C-like enum with 1 byte in size we could stuff 24 `ValType` in an array that results in the same size as a single `Vec<ValType>` and avoid an additional indirection. Maybe worth benchmarking if this is primarily about an optimization anyways.

Reply:
The [`ValType` type is about to become at least 4 bytes](https://github.com/bytecodealliance/wasm-tools/pull/701) I think and from there I think a `SmallVec`-style thing might still be useful but probably less so in the context of https://github.com/bytecodealliance/wasm-tools/issues/714. In any case I mostly wanted to just land a few easy wins without going too overboard on the complexity of the validator.